### PR TITLE
TW-1049 Notification improve iOS notification

### PR DESCRIPTION
**Pull request by drminh2807**
# Issue
- #1049 

# Demo

Cannot upload now, will upload later

### CODE REVIEW

File: lib/config/app_config.dart [hunk 1]
```diff
@@ -74,6 +74,10 @@ abstract class AppConfig {
   static const int thumbnailQuality = 70;
   static const int blurHashSize = 32;
   static const int imageQuality = 50;
+  static const String iOSKeychainSharingId =
+      'KUT463DS29.com.linagora.ios.twake.shared';
```
Comment (sherlockvn) [inline lib/config/app_config.dart:78]: what ?

File: lib/domain/model/keychain_sharing_restore_token.dart [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  Future saveToKeychain() =>
+      _secureStorage.write(key: session.userId, value: jsonEncode(toJson()));
+
+  static Future deleteFromKeychain({required String? userId}) {
+    if (userId != null) {
+      return _secureStorage.delete(key: userId);
+    } else {
+      return _secureStorage.deleteAll();
+    }
+  }
```
Comment (hoangdat) [inline lib/domain/model/keychain_sharing_restore_token.dart:43]: hope you don't forget to move it out of data class

File: lib/domain/model/keychain_sharing_session.dart [hunk 3]
```diff
@@ -0,0 +1,29 @@
+import 'package:json_annotation/json_annotation.dart';
+
+part 'keychain_sharing_session.g.dart';
+
+@JsonSerializable()
+class KeychainSharingSession {
+  String accessToken;
+  String? refreshToken;
+  String? oidcData;
```
Comment (hoangdat) [inline lib/domain/model/keychain_sharing_session.dart:9]: what is it?

File: lib/domain/model/keychain_sharing_session.dart [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  String? slidingSyncProxy;
+  String userId;
+  String deviceId;
```
Comment (hoangdat) [inline lib/domain/model/keychain_sharing_session.dart:12]: what kind of deviceId?

File: lib/domain/model/keychain_sharing_session.dart [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+part 'keychain_sharing_session.g.dart';
+
+@JsonSerializable()
+class KeychainSharingSession {
+  String accessToken;
+  String? refreshToken;
+  String? oidcData;
+  String? slidingSyncProxy;
+  String userId;
```
Comment (hoangdat) [inline lib/domain/model/keychain_sharing_session.dart:11]: you mean `matrixId`?

File: lib/utils/matrix_sdk_extensions/flutter_hive_collections_database.dart [hunk 6]
```diff
@@ -1,20 +1,25 @@
...
 import 'package:matrix/matrix.dart';
 import 'package:path_provider/path_provider.dart';
 
+import 'package:fluffychat/domain/model/keychain_sharing_restore_token.dart';
+import 'package:fluffychat/domain/model/keychain_sharing_session.dart';
+import 'package:fluffychat/utils/platform_infos.dart';
+import 'package:fluffychat/utils/storage_directory_utils.dart';
+
 class FlutterHiveCollectionsDatabase extends HiveCollectionsDatabase {
+  final String? userId;
```
Comment (hoangdat) [inline lib/utils/matrix_sdk_extensions/flutter_hive_collections_database.dart:17]: IMO it is a good solution, it can make our app can not support multiple account

File: lib/domain/model/keychain_sharing_restore_token.dart [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+  }) {
+    pusherNotificationClientIdentifier = session.userId.sha256Hash;
+  }
+
+  factory KeychainSharingRestoreToken.fromJson(Map<String, dynamic> json) =>
+      _$KeychainSharingRestoreTokenFromJson(json);
+
+  Map<String, dynamic> toJson() => _$KeychainSharingRestoreTokenToJson(this);
+
+  static FlutterSecureStorage get _secureStorage => const FlutterSecureStorage(
```
Comment (hoangdat) [inline lib/domain/model/keychain_sharing_restore_token.dart:27]: base on `FlutterSecureStorage` is very risky at the moment

**Comment by sherlockvn** [comment]
Can you summarize the steps to do this tickets :) because it's hard to review and follow

**Comment by drminh2807** [comment]
> Can you summarize the steps to do this tickets :) because it's hard to review and follow

please dont review *.swift files, I just copy 100% from Element X

File: scripts/patchs/ios-extension-debug.patch
[the remark this answers is not in the document]
Comment (drminh2807) [reply scripts/patchs/ios-extension-debug.patch:1]: you cannot debug NSE with FlutterAppDelegate
It retuns that crash
```
Cannot create a FlutterEngine instance in debug mode without Flutter tooling or Xcode.

To launch in debug mode in iOS 14+, run flutter run from Flutter tools, run from an IDE with a Flutter IDE plugin or run the iOS project from Xcode.
Alternatively profile and release mode apps can be launched from the home screen.
```
you should remove FlutterAppDelegate and use AppDelegate instead

File: lib/utils/background_push.dart [hunk 8]
```diff
@@ -205,6 +206,8 @@ class BackgroundPush {
                             "loc-args": [],
                           },
                         },
+                        "pusher_notification_client_identifier":
+                            client.pusherNotificationClientIdentifier,
```
Comment (hoangdat) [inline lib/utils/background_push.dart:210]: please add ADR for this PR and dont forget this. Thanks

File: lib/utils/matrix_sdk_extensions/flutter_hive_collections_database.dart (hunk 6 above)
Comment (drminh2807) [reply lib/utils/matrix_sdk_extensions/flutter_hive_collections_database.dart]: removed, should update later when multi account supported

File: lib/domain/model/keychain_sharing_session.dart (hunk 4 above)
Comment (drminh2807) [reply lib/domain/model/keychain_sharing_session.dart]: same with above

File: lib/domain/model/keychain_sharing_session.dart (hunk 5 above)
Comment (drminh2807) [reply lib/domain/model/keychain_sharing_session.dart]: same with above

File: lib/domain/model/keychain_sharing_restore_token.dart (hunk 7 above)
Comment (drminh2807) [reply lib/domain/model/keychain_sharing_restore_token.dart]: no need to care about it now

File: scripts/patchs/ios-extension-debug.patch
[the remark this answers is not in the document]
Comment (hoangdat) [reply scripts/patchs/ios-extension-debug.patch:1]: please also add it in ADR

File: ios/Runner/Runner.entitlements [hunk 9]
```diff
@@ -13,5 +13,9 @@
 	<array>
 		<string>group.com.linagora.ios.twake</string>
 	</array>
+	<key>keychain-access-groups</key>
+	<array>
+		<string>$(AppIdentifierPrefix)com.linagora.ios.twake.shared</string>
+	</array>
```
Comment (hoangdat) [inline ios/Runner/Runner.entitlements:19]: ADR

File: ios/Runner.xcworkspace/xcshareddata/swiftpm/Package.resolved
Comment (hoangdat) [inline ios/Runner.xcworkspace/xcshareddata/swiftpm/Package.resolved:1]: What is it?

File: ios/Runner.xcodeproj/xcshareddata/xcschemes/Runner.xcscheme [hunk 10]
```diff
@@ -38,8 +36,8 @@
             ReferencedContainer = "container:Runner.xcodeproj">
          </BuildableReference>
       </MacroExpansion>
-      <AdditionalOptions>
-      </AdditionalOptions>
+      <Testables>
+      </Testables>
```
Comment (hoangdat) [inline ios/Runner.xcodeproj/xcshareddata/xcschemes/Runner.xcscheme:40]: what is it?

File: ios/Runner/en.lproj/Localizable.strings
[the remark this answers is not in the document]
Comment (drminh2807) [reply ios/Runner/en.lproj/Localizable.strings:1]: actual no need, but I just copy all from Element X, it can make easy when you want copy it again from Element X in the future

File: ios/Runner.xcworkspace/xcshareddata/swiftpm/Package.resolved
Comment (drminh2807) [reply ios/Runner.xcworkspace/xcshareddata/swiftpm/Package.resolved:1]: it is Swift Package Manager, a cocoapod alternative

File: ios/Runner.xcodeproj/xcshareddata/xcschemes/Runner.xcscheme (hunk 10 above)
Comment (drminh2807) [reply ios/Runner.xcodeproj/xcshareddata/xcschemes/Runner.xcscheme:40]: it is changed by newly Xcode

File: lib/config/app_config.dart (hunk 1 above)
Comment (sherlockvn) [reply lib/config/app_config.dart:78]: do we need to have `KUT463DS29` prefix

Comment (drminh2807) [reply lib/config/app_config.dart:78]: yes 
https://developer.apple.com/documentation/security/keychain_services/keychain_items/sharing_access_to_keychain_items_among_a_collection_of_apps/#2974916

File: docs/adr/0012-improve-ios-notification.md [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+The motivation behind this decision is the inability to execute Dart code for decrypting notifications when the iOS app is running in the background. This limitation is different from Android or Web, where Dart code can easily run in such scenarios.
+
+## Decision
+We have decided to make the following changes to address the mentioned issue:
+
+1. **Notification Service Extension (NSE):**
+   - Utilize the [Notification Service Extension](https://developer.apple.com/documentation/usernotifications/modifying_content_in_newly_delivered_notifications) provided by Apple to modify the content of notifications.
+
+2. **MatrixRustSDK Integration:**
+   - Integrate MatrixRustSDK to decrypt messages within the Notification Service Extension.
```
Comment (hoangdat) [inline docs/adr/0012-improve-ios-notification.md:19]: What is the version, how to manage (update/change) the SDK?

File: docs/adr/0012-improve-ios-notification.md [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+3. **Automated Script for Source Code Copy:**
+   - Develop a script written in Node.js to automate the process of copying the entire source code related to the Notification Service Extension from Element X, a known version.
+
+4. **Dart Code Adjustment:**
+   - Modify Dart code to ensure compatibility with the Notification Service Extension from Element X.
+
+5. **Debugging Patch with AppDelegate:**
+   - Add the patch `scripts/patchs/ios-extension-debug.patch` to address debugging issues. This patch removes FlutterAppDelegate and suggests using AppDelegate instead for smoother debugging during Notification Service Extension development.
+
+6. **Add pusher_notification_client_identifier to Default Payload:**
```
Comment (hoangdat) [inline docs/adr/0012-improve-ios-notification.md:30]: please add the example

File: docs/adr/0012-improve-ios-notification.md [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+
+7. **Keychain Access Groups:**
+   - Utilize `keychain-access-groups` to enable data exchange between the Notification Service Extension (NSE) and the main app by using a shared Keychain Access Group. This ensures security and safe data transmission (See details at [link](https://developer.apple.com/documentation/security/keychain_services/keychain_items/sharing_access_to_keychain_items_among_a_collection_of_apps/)).
+
+8. **INSendMessageIntent:**
+   - Use `INSendMessageIntent` to support displaying profile pictures in notifications. Instead of using regular notifications, integrating `INSendMessageIntent` ensures that the sender's image is displayed correctly in the notification (See details at [link](https://stackoverflow.com/questions/68198452/ios-15-communication-notification-picture-not-shown)).
+
+9. **iOS Version Support Below 16:**
+   - Communicate that Element X's Notification Service Extension supports iOS 16 and above. Therefore, Twake NSE will also have a similar requirement. Users with iOS versions below 16 can still install Twake, but they won't be able to decrypt notifications.
```
[the remark this answers is not in the document]
Comment (drminh2807) [reply docs/adr/0012-improve-ios-notification.md:40]: old element require start from iOS 14

**Review by hoangdat — APPROVED** [review]
